### ISSUE TITLE
[bug] ZIO.raceFirst sometimes prevents scope finalisers from running

### ISSUE THREAD

**Issue by myazinn** [zio/zio#9203]
I tried ZIO 2.0.0, 2.1.0, 2.1.9 (latest). All of them have the same issue.

Here's a test that reproduces the error
```scala
import zio._
import zio.test.TestAspect.timeout
import zio.test._

import java.util.concurrent.atomic.AtomicBoolean

object ZIORaceFirstBug extends ZIOSpecDefault {
  override def spec: Spec[Any, Any] =
    suite("ZIO.raceFirst")(
      test("should always call finalizers") {
        for {
          isRunning <- ZIO.succeed(new AtomicBoolean(true))

          backgroundBlockingStuff = ZIO.attemptBlocking(while (isRunning.get()) {})

          acquire = backgroundBlockingStuff.fork
          release = ZIO.succeed(isRunning.set(false))

          eff = ZIO.acquireRelease(acquire)(_ => release)

          // _ <- ZIO.scoped(eff) // this one works
          _ <- ZIO.scoped(ZIO.raceFirst(eff, Nil))

        } yield assertCompletes
      } @@ timeout(10.seconds)
    )
}

```

For some reason, when there's long-running blocking background operation wrapped into `ZIO.raceFirst`, finalisers are not called and test terminates due to timeout. If I replace `ZIO.scoped(ZIO.raceFirst(eff, Nil))` with `ZIO.scoped(eff)`, test immediately passes.
IMO racing an effect against an empty list should have no impact on how it behaves at all. My real case is way more complicated and I'm actually racing against a bunch of other stuff, so simple replacement as in this test is not an option. As a workaround I can do `(effect +: effects).reduce(_.raceFirst(_)` and it seems to work fine, but it's still surprising that `ZIO.raceFirst` doesn't work.

**Comment by narma** [comment]
If you change `acquire = backgroundBlockingStuff.fork` to  `acquire = backgroundBlockingStuff.forkScoped` this test will start passing again.

**Comment by myazinn** [comment]
@narma that's interesting 🤔 I wonder why

**Comment by kyri-petrou** [comment]
> @narma that's interesting 🤔 I wonder why

I believe the reason for that is that `forkScoped` adds the finalizers to the Scope seen by the parent fiber. Whereas when you use `.fork`, the finalizers are added to the scope of the child fiber. I'm not sure if this is a bug or a feature, but `inheritAll` (which race uses) doesn't inherit the scope of child fibers. Perhaps it should? I'm not sure what the right answer is here to be honest.

**Comment by myazinn** [comment]
@kyri-petrou well, I'd say `ZIO.raceFirst(eff, Nil)` should always behave exactly like simple `eff` (even though that's not my real case, and optimising for this specific case wouldn't fix my issue)
And subsequently `ZIO.scoped(ZIO.raceFirst(eff, Nil))` should behave just like `ZIO.scoped(eff)`.

**Comment by jdegoes** [comment]
/bounty $350

It is indeed true that `ZIO.raceFirst(eff, Nil)` should behave exactly like `eff`, for all `eff`.

**Comment by Saturn225** [comment]
/attempt #9203

<div id="algora-attempt" />

| [Algora profile](https://console.algora.io/@/SankeerthKoraboina) | Completed bounties | Tech | Active attempts | Options |
| --- | --- | --- | --- | --- |
| @Saturn225 | 2 bounties from 2 projects | <div align="center">Rust, Scala, <br />HTML</div> | <div align="center"></div> | [Cancel attempt](https://console.algora.io/api/bounties/cm3926yok000kl703t8t4vq8k/cancel-attempt) |

**Comment by regiskuckaertz** [comment]
/attempt #9203

<div id="algora-attempt" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/cm3926yok000kl703t8t4vq8k/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by kyri-petrou** [comment]
Just want to add that the same issue exists with `ZIO.race` / `ZIO.raceFirst` and should ideally be addressed as part of this bounty. e.g., changing this line in the reproducer above `ZIO.scoped(ZIO.raceFirst(eff, Nil))` to one of the following current hangs:

- `ZIO.race(eff, ZIO.never)`
- `ZIO.raceFirst(eff, ZIO.never)`

**Comment by digy** [comment]
Issue was solved by #9353.
We can close it here

**Comment by kyri-petrou** [comment]
Closed by https://github.com/zio/zio/pull/9353
